tests/cpp/engine/threaded_engine_test.cc

```diff
@@ -257,49 +257,80 @@ TEST(Engine, PushFunc) {
 
 TEST(Engine, PushFuncND) {
   auto ctx = mxnet::Context{};
-  mxnet::NDArray nd(ctx);
-
-  // Test #1
-  LOG(INFO) << "===== Test #1: PushAsyncND param and deleter =====";
-  int* a = new int(100);
-  int res = MXEnginePushAsyncND(FooAsyncFunc, a, FooFuncDeleter, &ctx, &nd, 1, nullptr, 0);
-  EXPECT_EQ(res, 0);
-
-  // Test #2
-  LOG(INFO) << "===== Test #2: PushAsyncND NULL param and NULL deleter =====";
-  res = MXEnginePushAsyncND(FooAsyncFunc, nullptr, nullptr, &ctx, nullptr, 0, &nd, 0);
-  EXPECT_EQ(res, 0);
-
-  // Test #3
-  LOG(INFO) << "===== Test #3: PushAsyncND invalid number of const nds =====";
-  res = MXEnginePushAsyncND(FooAsyncFunc, nullptr, nullptr, &ctx, &nd, -1, nullptr, 0);
-  EXPECT_EQ(res, -1);
-
-  // Test #4
-  LOG(INFO) << "===== Test #4: PushAsyncND invalid number of mutable nds =====";
-  res = MXEnginePushAsyncND(FooAsyncFunc, nullptr, nullptr, &ctx, nullptr, 0, &nd, -1);
-  EXPECT_EQ(res, -1);
-
-  // Test #5
-  LOG(INFO) << "===== Test #5: PushSyncND param and deleter =====";
-  int* b = new int(101);
-  res = MXEnginePushSyncND(FooSyncFunc, b, FooFuncDeleter, &ctx, &nd, 1, nullptr, 0);
-  EXPECT_EQ(res, 0);
-
-  // Test #6
-  LOG(INFO) << "===== Test #6: PushSyncND NULL param and NULL deleter =====";
-  res = MXEnginePushSyncND(FooSyncFunc, nullptr, nullptr, &ctx, nullptr, 0, &nd, 1);
-  EXPECT_EQ(res, 0);
-
-  // Test #7
-  LOG(INFO) << "===== Test #7: PushSyncND invalid number of const nds =====";
-  res = MXEnginePushSyncND(FooSyncFunc, nullptr, nullptr, &ctx, &nd, -1, nullptr, 0);
-  EXPECT_EQ(res, -1);
-
-  // Test #8
-  LOG(INFO) << "===== Test #8: PushSyncND invalid number of mutable nds =====";
-  res = MXEnginePushSyncND(FooSyncFunc, nullptr, nullptr, &ctx, nullptr, 0, &nd, -1);
-  EXPECT_EQ(res, -1);
+  std::vector<mxnet::NDArray*> nds;
```

Comment:
In fact, I'm thinking if it's more appropriate to use
```c++
std::vector<mxnet::NDArray> nds;
```
Do we have to use `std::vector<mxnet::NDArray*>`?

Reply:
We have to use `std::vector<mxnet::NDArray*>`, because the type of the argument of the two APIs is an array pointer of `NDArray*`. Besides, it avoids the potential copy of NDArray.

Reply:
There is an argument Context in the constructor of NDArray. I do not know how to use `std::vector<mxnet::NDArray>`.

Reply:
Yes, we have to use `std::vector<NDArray*>` if we keep the interface to be `NDArrayHandle *`. However, I'm thinking whether we could directly use the `std::vector<mxnet::NDArray> vec;` and use `nds.emplace_back(mxnet::NDArray(ctx))` or `nds.push_back(std::move(temp_arr))` to fill the vector. In that case, the existing API will not be changed.

Reply:
It sounds good, but I am worry that the type of arguments (`const_vars_handle` and `mutable_vars_handle`) of the two APIs [`MXEnginePushAsync`](https://github.com/apache/incubator-mxnet/blob/master/src/c_api/c_api.cc#L1466) and [`MXEnginePushSync`](https://github.com/apache/incubator-mxnet/blob/master/src/c_api/c_api.cc#L1507) is `EngineVarHandle`, namely `void*`. It casts `void*` to `VarHandle*`, namely `Var**` in https://github.com/apache/incubator-mxnet/blob/master/src/c_api/c_api.cc#L1475. Therefore, I don't know how to decide the type of `const_nds_handle` and `mutable_nds_handle` in `MXEnginePushAsyncND` and `MXEnginePushSyncND`.

Reply:
To keep the consistency, `const_var_handle` and `mutable_var_handle` is the pointer of an array of `VarHandle`, and `const_nds_handle` and `mutable_var_handle` is the pointer of an array of `NDArrayHandle`.

Reply:
I now understand the logic here. To make the API consistent, I think we should also change the interface of `MXEnginePushAsync` and `MXEnginePushSync`. We should be safe to replace `EngineVarHandle` with `VarHandle*`. Am I right here ? @apeforest @yuxihu